rust/src/modbus/detect.rs

```diff
@@ -0,0 +1,563 @@
+/* Copyright (C) 2021 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::modbus::ModbusTransaction;
+use lazy_static::lazy_static;
+use regex::Regex;
+use sawp_modbus::{AccessType, CodeCategory, Data, Flags, FunctionCode, Message};
+use std::ffi::CStr;
+use std::ops::{Range, RangeInclusive};
+use std::os::raw::{c_char, c_void};
+use std::str::FromStr;
+
+lazy_static! {
+    static ref ACCESS_RE: Regex = Regex::new(
+        "^\\s*\"?\\s*access\\s*(read|write)\
+        \\s*(discretes|coils|input|holding)?\
+        (?:,\\s*address\\s+([<>]?\\d+)(?:<>(\\d+))?\
+        (?:,\\s*value\\s+([<>]?\\d+)(?:<>(\\d+))?)?)?\
+        \\s*\"?\\s*$"
+    )
+    .unwrap();
+    static ref FUNC_RE: Regex = Regex::new(
+        "^\\s*\"?\\s*function\\s*(!?[A-z0-9]+)\
+        (?:,\\s*subfunction\\s+(\\d+))?\\s*\"?\\s*$"
+    )
+    .unwrap();
+    static ref UNIT_RE: Regex = Regex::new(
+        "^\\s*\"?\\s*unit\\s+([<>]?\\d+)\
+        (?:<>(\\d+))?(?:,\\s*(.*))?\\s*\"?\\s*$"
+    )
+    .unwrap();
+}
+
+#[derive(Debug, PartialEq)]
+pub struct DetectModbusRust {
```

Comment:
Should this be more like a C union (a Rust enum) as we cannot have the whole fields set at one time ?
(it is either units, or function ...)

Reply:
I think we can have more than one field set according to the [modbus keyword documentation](https://suricata.readthedocs.io/en/suricata-6.0.2/rules/modbus-keyword.html?highlight=modbus). This is why we chose a `struct`.
```
modbus: unit 10<>20, access write holding, address 500, value >200    
```

Reply:
Ok.

But then `rs_modbus_inspect` seems to do too early `return` like `return inspect_data(msg, modbus) as u8;`
Cannot we have `modbus.access_type` and `modbus.category` set ?

Reply:
This functionality was ported from the C parser and refactored to use early returns instead of nested ifs but it should have the same effect.

See [detect-engine-modbus.c#L199](https://github.com/OISF/suricata/blob/master/src/detect-engine-modbus.c#L199)

Reply:
The C behavior looks confusing to me.

Is it possible to have a signature like `alert modbus any any -> any any (msg:"Modbus function number test"; modbus: function 1, access read coils; sid:1; rev:1;)` ?
If not, why ?

Reply:
I think the documentation is a bit more clear than the C code: https://suricata.readthedocs.io/en/suricata-6.0.0/rules/modbus-keyword.html

The function and access keywords cannot be used together, probably because one is more specific than the other. I.e. `function 1` is more specific than `access read coils`.

Reply:
Ok thanks :-)